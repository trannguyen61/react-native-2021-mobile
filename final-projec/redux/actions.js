import { fetchMovies } from '../api/movies';
import { login } from '../api/auth'

export const SEARCH_QUERY_SENT = 'SEARCH_QUERY_SENT';
export const NEW_RESULTS_RECEIVED = 'NEW_RESULTS_RECEIVED';
export const ADDITIONAL_RESULTS_RECEIVED = 'ADDITIONAL_RESULTS_RECEIVED';
export const RESULTS_ERROR = 'RESULTS_ERROR';
export const LOGIN = 'LOGIN'
export const LOGIN_FAILED = 'LOGIN_FAILED'
export const LOGOUT = 'LOGOUT'

export const onLogin = (username, password, loginFn = login) => async (dispatch) => {
  const response = loginFn(username, password)
  if (response.token) {
    dispatch({
      type: LOGIN,
      payload: { token: response.token }
    })
  } else {
    dispatch({
      type: LOGIN_FAILED,
      payload: { loginErr: response.err }
    })
  }
}

export const onFetchMovies = (search, pageNumber, fetchFn = fetchMovies) => async (dispatch) => {
  dispatch({ type: SEARCH_QUERY_SENT });
  try {
    const response = await fetchFn(search, pageNumber);
    const results = response.results;

    const totalResults = response.totalResults;
    const maxPages = Math.ceil(totalResults / 10); // 10 results per page

    if (pageNumber === 1) {
      dispatch({
        type: NEW_RESULTS_RECEIVED,
        payload: { movies: results, maxPages: maxPages },
      });
    } else {
      dispatch({
        type: ADDITIONAL_RESULTS_RECEIVED,
        payload: { movies: results },
      });
    }
  } catch (err) {
    dispatch({ type: RESULTS_ERROR, payload: {
        err: 'There\'s something wrong with the server. Please try again.'
    } });
  }
};