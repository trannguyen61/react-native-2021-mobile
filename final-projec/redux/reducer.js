import { combineReducers } from 'redux';
import {
  SEARCH_QUERY_SENT,
  NEW_RESULTS_RECEIVED,
  ADDITIONAL_RESULTS_RECEIVED,
  RESULTS_ERROR,
  CLEAR_RESULTS,
  LOGIN,
  LOGIN_FAILED,
  LOGOUT
} from './actions';

const merge = (prev, next) => Object.assign({}, prev, next);

const authReducer = (state = {}, action) => {
  switch (action.type) {
    case LOGIN:
      return merge(state, {
        token: action.payload.token
      })
    case LOGIN_FAILED: 
      return merge(state, {
        loginErr: action.payload.loginErr
      })
    case LOGOUT: 
      return merge(state, {
        token: null
      })
    default:
      return state
  }
}

const resultsReducer = (state = {}, action) => {
  switch (action.type) {
    case NEW_RESULTS_RECEIVED:
      return merge(state, {
        movies: action.payload.movies,
        maxPages: action.payload.maxPages,
      });
    case ADDITIONAL_RESULTS_RECEIVED:
      return merge(state, {movies: [...state.movies, ...action.payload.movies]} );
    case RESULTS_ERROR:
      return merge(state, { resultsErr: action.payload.err });
    default:
      return state;
  }
};

const reducer = combineReducers({
  results: resultsReducer,
  auth: authReducer
});

export default reducer;