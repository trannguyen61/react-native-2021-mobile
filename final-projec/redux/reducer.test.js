import reducer from "./reducer";
import * as actions from "./actions";

const DEFAULT_STATE = {
  auth: {
    token: null,
    loginErr: null,
  },
  results: {
    movies: null,
    maxPages: null,
    resultsErr: null,
  },
};

const ADDITIONAL_MOVIES_STATE = {
    auth: {
      token: null,
      loginErr: null,
    },
    results: {
      movies: [],
      maxPages: 2,
      resultsErr: null,
    },
};

const merge = (prev, next) => Object.assign({}, prev, next);

describe("auth reducer", () => {
    it("should return the initial state", () => {
      expect(reducer(DEFAULT_STATE, {})).toEqual(DEFAULT_STATE);
    });
  
    it("should handle LOGIN", () => {
      const action = {
        type: actions.LOGIN,
        payload: { token: 'a' },
      };
      expect(reducer(DEFAULT_STATE, action)).toEqual(
        merge(DEFAULT_STATE, {
          auth: merge(action.payload, { loginErr: null }),
        })
      );
    });
  
    it("should handle LOGIN_FAILED", () => {
      const action = {
        type: actions.LOGIN_FAILED,
        payload: { loginErr: 'Error' },
      };
      expect(reducer(DEFAULT_STATE, action)).toEqual(
        merge(DEFAULT_STATE, {
          auth: merge({loginErr: action.payload.loginErr}, { token: null }),
        })
      );
    });
  });

describe("results reducer", () => {
  it("should return the initial state", () => {
    expect(reducer(DEFAULT_STATE, {})).toEqual(DEFAULT_STATE);
  });

  it("should handle NEW_RESULTS_RECEIVED", () => {
    const action = {
      type: actions.NEW_RESULTS_RECEIVED,
      payload: { movies: [], maxPages: 1 },
    };
    expect(reducer(DEFAULT_STATE, action)).toEqual(
      merge(DEFAULT_STATE, {
        results: merge(action.payload, { resultsErr: null }),
      })
    );
  });

  it("should handle RESULTS_ERROR", () => {
    const action = {
      type: actions.RESULTS_ERROR,
      payload: { err: 'There\'s something wrong with the server. Please try again.' },
    };
    expect(reducer(DEFAULT_STATE, action)).toEqual(
      merge(DEFAULT_STATE, {
        results: merge({resultsErr: action.payload.err}, { movies: null, maxPages: null }),
      })
    );
  });

  it("should handle ADDITIONAL_RESULTS_RECEIVED", () => {
    const action = {
      type: actions.ADDITIONAL_RESULTS_RECEIVED,
      payload: { movies: [] },
    };
    expect(reducer(ADDITIONAL_MOVIES_STATE, action)).toEqual(
      merge(ADDITIONAL_MOVIES_STATE, {
        results: merge(action.payload, { resultsErr: null, maxPages: 2 }),
      })
    );
  });
});
