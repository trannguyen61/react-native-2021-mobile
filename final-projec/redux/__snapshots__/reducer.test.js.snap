// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`results reducer receives new movies 1`] = `
Object {
  "auth": Object {
    "loginErr": null,
    "token": null,
  },
  "results": Object {
    "maxPages": null,
    "movies": null,
    "resultsErr": null,
  },
}
`;
